WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SoundRecorder.kt

```diff
@@ -0,0 +1,159 @@
+        IDLE, RECORDING, PLAYING
+    }
+
+    /**
+     * Plays the recorded file, if any.
+     *
+     * Returns when playing the file is finished.
+     *
+     * This is cancellable, and cancelling it will stop playback.
+     */
+    suspend fun play() {
+        if (state != State.IDLE) {
+            Log.w(TAG, "Requesting to play while state was not IDLE")
+            return
+        }
+
+        // Check if there isn't a recording to play
+        if (!File(context.filesDir, outputFileName).exists()) return
+
+        state = State.PLAYING
+
+        val intSize = AudioTrack.getMinBufferSize(RECORDING_RATE, CHANNELS_OUT, FORMAT)
+
+        val audioTrack = AudioTrack.Builder()
+            .setAudioAttributes(
+                AudioAttributes.Builder()
+                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
+                    .setUsage(AudioAttributes.USAGE_MEDIA)
+                    .build()
+            )
+            .setBufferSizeInBytes(intSize)
+            .setAudioFormat(
+                AudioFormat.Builder()
+                    .setSampleRate(RECORDING_RATE)
+                    .setChannelMask(CHANNELS_OUT)
+                    .setEncoding(FORMAT)
+                    .build()
+            )
+            .setTransferMode(AudioTrack.MODE_STREAM)
+            .build()
+
+        val buffer = ByteArray(intSize * 2)
+
+        audioTrack.setVolume(AudioTrack.getMaxVolume())
+        audioTrack.play()
+
+        try {
+            withContext(Dispatchers.IO) {
+                context.openFileInput(outputFileName).buffered().use { bufferedInputStream ->
+                    while (isActive) {
+                        val read = bufferedInputStream.read(buffer, 0, buffer.size)
+                        if (read < 0) break
+                        audioTrack.write(buffer, 0, read)
+                    }
+                }
+            }
+        } finally {
+            audioTrack.release()
+            state = State.IDLE
+        }
+    }
+
+    /**
+     * Records from the microphone.
+     *
+     * This method is cancellable, and cancelling it will stop recording.
+     */
+    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
+    suspend fun record() {
+        if (state != State.IDLE) {
+            Log.w(TAG, "Requesting to start recording while state was not IDLE")
+            return
+        }
+
+        state = State.RECORDING
+
+        val intSize = AudioRecord.getMinBufferSize(RECORDING_RATE, CHANNEL_IN, FORMAT)
+
+        val audioRecord = AudioRecord.Builder()
+            .setAudioSource(MediaRecorder.AudioSource.MIC)
+            .setAudioFormat(
+                AudioFormat.Builder()
+                    .setSampleRate(RECORDING_RATE)
+                    .setChannelMask(CHANNEL_IN)
+                    .setEncoding(FORMAT)
+                    .build()
+            )
+            .setBufferSizeInBytes(intSize * 3)
+            .build()
+
+        val buffer = ByteArray(intSize)
```

Comment:
nit: Maybe move this into the `withContext` block, just so its closer to where it's being used.

Reply:
Done for both cases in dbfd80d!